### ISSUE TITLE
Brew installed version doesn't work with plugins

### ISSUE THREAD

**Issue by rshurts** [httpie/httpie#566]
Brew is listed as the recommended way of installing httpie on macOS. However, it doesn't work with auth plugins.

For example, if you `pip3 install requests-hawk` and then run `http --help` hawk will not show as an auth type. If you `pip3 install httpie-oauth` it will install httpie via pip as a dependency and overwrite the brew installed link in `/usr/local/bin/http` and now all the plugins will show because it isn't using the brew installed version.

I suggest changing the documentation to read `pip3 install httpie` as the recommended method of installing on macOS.

**Comment by jkbrzt** [comment]
@rshurts you're right. Plugin installation for Homebrew-based installs doesn't work out of the box. 

There are some advantages to keeping Homebrew as the recommended method, though. Mainly it makes updates really easy and the package can depend on Python 3 (i.e. much better SSL support, etc.). 

The ideal world solution would be to make all plugins available as Homebrew packages (e.g. `brew install httpie-oauth`, etc.) . But that would mean additional work for maintainers so not all plugins would end up in Homebrew. 

Using `pip` to install plugins looks like the easier way to go. To make it work with Hombrew, it would mean instructing `pip` to use  (currently) `/usr/local/Cellar/httpie/0.9.8_2/libexec/lib/python3.6/site-packages` as the installation directory. 

But plugins installed there won't survive HTTPie upgrades. 

Perhaps it would be possible to tweak the [Homebrew formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/httpie.rb) to make it possible to load installed plugins from the Homebrew Python 3 `site-packages` (or from another location). 

…or some sort of a custom plugin installer.

**Comment by rshurts** [comment]
In the meantime, maybe a note in the documentation around plugins saying "brew doesn't work with plugins and to pip install httpie for plugin functionality" would save some headaches.

**Comment by aheissenberger** [comment]
There is also a problem with the read rights when user has no admin rights:
```bash
$ http --help
Traceback (most recent call last):
  File "/usr/local/bin/http", line 6, in <module>
    from pkg_resources import load_entry_point
  File "/usr/local/Cellar/httpie/1.0.0/libexec/lib/python3.7/site-packages/pkg_resources/__init__.py", line 3123, in <module>
    @_call_aside
  File "/usr/local/Cellar/httpie/1.0.0/libexec/lib/python3.7/site-packages/pkg_resources/__init__.py", line 3107, in _call_aside
    f(*args, **kwargs)
  File "/usr/local/Cellar/httpie/1.0.0/libexec/lib/python3.7/site-packages/pkg_resources/__init__.py", line 3136, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/usr/local/Cellar/httpie/1.0.0/libexec/lib/python3.7/site-packages/pkg_resources/__init__.py", line 569, in _build_master
    ws = cls()
  File "/usr/local/Cellar/httpie/1.0.0/libexec/lib/python3.7/site-packages/pkg_resources/__init__.py", line 562, in __init__
    self.add_entry(entry)
  File "/usr/local/Cellar/httpie/1.0.0/libexec/lib/python3.7/site-packages/pkg_resources/__init__.py", line 618, in add_entry
    for dist in find_distributions(entry, True):
  File "/usr/local/Cellar/httpie/1.0.0/libexec/lib/python3.7/site-packages/pkg_resources/__init__.py", line 1965, in find_on_path
    for dist in factory(fullpath):
  File "/usr/local/Cellar/httpie/1.0.0/libexec/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2035, in distributions_from_metadata
    root, entry, metadata, precedence=DEVELOP_DIST,
  File "/usr/local/Cellar/httpie/1.0.0/libexec/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2492, in from_location
    py_version=py_version, platform=platform, **kw
  File "/usr/local/Cellar/httpie/1.0.0/libexec/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2867, in _reload_version
    md_version = _version_from_file(self._get_metadata(self.PKG_INFO))
  File "/usr/local/Cellar/httpie/1.0.0/libexec/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2456, in _version_from_file
    line = next(iter(version_lines), '')
  File "/usr/local/Cellar/httpie/1.0.0/libexec/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2646, in _get_metadata
    for line in self.get_metadata_lines(name):
  File "/usr/local/Cellar/httpie/1.0.0/libexec/lib/python3.7/site-packages/pkg_resources/__init__.py", line 1411, in get_metadata_lines
    return yield_lines(self.get_metadata(name))
  File "/usr/local/Cellar/httpie/1.0.0/libexec/lib/python3.7/site-packages/pkg_resources/__init__.py", line 1407, in get_metadata
    value = self._get(self._fn(self.egg_info, name))
  File "/usr/local/Cellar/httpie/1.0.0/libexec/lib/python3.7/site-packages/pkg_resources/__init__.py", line 1519, in _get
    with open(path, 'rb') as stream:
PermissionError: [Errno 13] Permission denied: '/usr/local/Cellar/httpie/1.0.0/libexec/lib/python3.7/site-packages/idna-2.7-py3.7.egg-info/PKG-INFO'
```

**I was able to fix it with** 
```sh
chmod +r /usr/local/Cellar/httpie/1.0.0/libexec/lib/python3.7/site-packages/idna-2.7-py3.7.egg-info/PKG-INFO
```

but switched to `pip3 install httpie` because of it made it easier to install plugins.

**Comment by sensibleish** [comment]
Since I upgraded to 10.14 Mojave I'm having the permissions error mentioned by @aheissenberger above. I have the same setup where I install homebrew using a local admin account but do normal operation in a standard account. I'm not a Python person and don't understand this discussion of plugins. I do not recall trying to add anything httpie beyond whatever comes as part of the homebrew formula. `pip3 list` only shows 3 items which presumably were part of the initial Python bundle.

Is this fundamentally a Python3 problem (at least, Python as installed via homebrew)? Should I be reporting it to the maintainers?

**Comment by aariafernandes** [comment]
[Q1 and ANS1.docx](https://github.com/jakubroztocil/httpie/files/3488465/Q1.and.ANS1.docx)

**Comment by aariafernandes** [comment]
dddddddddddd

**Comment by aariafernandes** [comment]
[Presentation.pdf](https://github.com/jakubroztocil/httpie/files/3488467/Presentation.pdf)

**Comment by ykdui** [comment]
Buy Valid Cvv CC Dumps Track 1/2 CC SSN DOB Track-1/2-FULLZ-Transfer/Western union Bank-Log-In-PayPal-Transfer-and-Log-In Buy/sell CVV CC DOB SSN FULLZ - Dumps Pin...
Telegram.@Ykidven

Cc Format Is
Format is:

Card Number | Exp. Date | CVV/CVV2 | First Name | Last Name | Street | City 
| State | Zip Code | Country | Phone | Type Of Card | Bank Name

I'm Loading All Types Of Paypal Account With No ChargeBacks... Cash app Loading I Can Load All Non Verified Account with No Problem Or Verified PayPal Account..I Even Sell Paypal Logs At Cheap Prices....Pm Me If You Interested In Loading Your Paypal Account
Telegram.@Ykidven

Paypal Transfer Price List
20$ for 250$
30$ for 350$
40$ for 400$
50$ for 500$
60$ for 600$
70$ for 700$
80$ for 800$
90$ for 950$
100$ for 1500$


CashApp Transfer Price List
20$ for 250$
30$ for 350$
40$ for 400$
50$ for 500$
60$ for 600$
70$ for 700$
80$ for 800$
90$ for 950$
100$ for 1500$
200$ for 3000$

Telegram@Ykidven

Doing Wu Transfers
I Am Doing Wu Transfer Here And I Need I Very Serious Buyer Who Also Want To Work With Me And Make Money With Me Here...
Note...i will Do The Transfer Then After you Track The Mtcn You Make Payment Before you have the sender's Info

Western Union Price List
30$ for 350$
40$ for 400$
50$ for 500$
60$ for 600$
70$ for 700$
80$ for 800$
90$ for 950$
100$ for 1500$
200$ for 2500$

Telegram.@Ykidven
Telegram.@Ykidven
Telegram.@Ykidven

ICQ...clarity.baron@bk.ru
ICQ...clarity.baron@bk.ru
ICQ...clarity.baron@bk.ru

**Comment by nicbet** [comment]
For what it's worth and perhaps other users googling for an answer, installing a plugin worked fine for me like this:

```
pip3 install -t /usr/local/Cellar/httpie/2.0.0/libexec/lib/python3.8/site-packages httpie-jwt-auth
```

Thanks https://github.com/jakubroztocil/httpie/issues/566#issuecomment-285210693

**Comment by matthew16550** [comment]
Just thinking out loud here, I haven't actually tried this.

Perhaps at runtime we add `~/.httpie/plugins/` to the search path (if it exists) and recommend Mac users do this:

`pip install -t ~/.httpie/plugins httpie-jwt-auth`

Suspect it might work fine for 80-90% of the use cases but no idea what might be in that last 10% !

**Comment by BoboTiG** [comment]
It is preferable rather my proposal on #1108 IMO.

Still:

- What would happen if someone installed a plugin in the folder using a different Python version (like Python 2.7) when HTTPie is running Python 3.9?
- Plugin requires HTTPie to be installed, so they install it, making the folder bigger (that's an issue on plugin themselves and not really a problem I guess), and having 2 `httpie` modules available.

All those solutions are not ideal, I'll see how other projects are doing. If you have examples, I would be glad to check them :)

**Comment by matthew16550** [comment]
> Plugin requires HTTPie to be installed, so they install it, making the folder bigger 

Good point, my random sampling of 4 plugins finds they all require `httpie` as an install dependency.

**Comment by matthew16550** [comment]
I've done a little work on [Nikola](https://getnikola.com/), it manages plugin installs itself.  Running `nikola plugin -i foo` will download and extract a foo.zip, then [run pip as a subprocess](https://github.com/getnikola/nikola/blob/7e03f76d8c0b9c05ffbb738f3c3c100da76eeb99/nikola/plugins/command/plugin.py#L254) to install any dependencies.

Maybe we could borrow the second half of that idea and do `http --install-plugin foo` to launch `pip` as a subprocess with the right settings to install the plugin next to the `httpie` module so in Homebrew installs it would end up somewhere like `/usr/local/Cellar/httpie/2.3.0/libexec/lib/python3.9/site-packages/foo`.

### Pros
* Plugins will be installed using the same version of Python that Httpie is using.
* No duplicate installation of `httpie`.
* Does not add plugins to a global dir (e.g. `/usr/local/lib/python3.7/site-packages`).
* No worrying about which of the many pips on my machine I should use to install the plugin.

### Cons
* I think plugins will need to be manually reinstalled every time Homebrew updates Httpie to a new version.
* Needing to reinstall the plugins will be quite annoying!

Kind of suspect it would go against Homebrew philosophy if we modify a dir managed by Homebrew in this way but it's difficult to see a clean solution to all the complications here.

I'm not convinced this is a good idea but maybe it's a step nearer to one.

**Comment by matthew16550** [comment]
Another example: For [JMeter](https://jmeter.apache.org/), [this blog](https://medium.com/@sdanerib/run-jmeter-with-plugins-in-macos-8a6654fc0b38) recommends manually copying plugins into the Homebrew dir.

**Comment by isidentical** [comment]
I focused on this today and seems like it is finally working (install/uninstall/list). Here is the draft branch to give an impression of the implementation, though it is nowhere near production-ready (in terms of UI): https://github.com/httpie/httpie/compare/master...isidentical:httpie-plugins?expand=1. There were a couple of caveats (e.g pip doesn't support uninstallation from a custom scheme etc) but they all seem to work now. The next steps would be creating a proper UI for this (especially in terms of error handling), and figuring out a way to write tests.

**Comment by isidentical** [comment]
We have finally landed `httpie plugins` interface, though the initial QA on brew is not yet completed. I'll probably try to complete it within today or tomorrow and then close this ticket.

**Comment by isidentical** [comment]
🎉 This feature is now publicly available in [HTTPie for Terminal 3.0](https://httpie.io/blog/httpie-3.0.0):

👉🏻  https://httpie.io/docs/cli/plugin-manager
